build api commands based on API FreeIPA json api spec

schlitzered opened the pull request:
Hi there,

i rebuild the client commands bases on the API FreeIPA api spec, so now basically all API endpoints are supported.

i did this using a tool i created for this called "py_ipa_api_recreated". (can be found in contrib folder)

currently it will consume the JSON API spec, that has to be fetched first. and create a file called "meta_api.py"

this file contains a class with:
-  all api endpoints as a function,
-  all known params, and default values
-  api doc

the Client class will subclass fomr the MetaAPI Class, found in meta_api.py.

because of auto creation, i was not able to keep parameter names of the different functions :-/

let me know what you think, if you like it, i will invest some more time, and maybe also add parameter valaidation, based on the JSON API spec.

schlitzered committed:
build api commands based on API FreeIPA json api spec

AmbientLighter commented:
Thank you very much for your contribution. I believe that the approach you have implemented is the way to go. However I am concerned about backward compatibility since argument names are different now.

schlitzered commented:
if you approve the general direction, how about a 1.x.x release, and making the old api available as "ClientLegacy"?

AmbientLighter commented:
Yes, it sounds good to me.

schlitzered commented:
ok nice, i am currently using the "name" attribute for parameter names, but there is also a attribute called "cli_name" that sometimes has a more friendly name. 

for example the "add_user" method has the attribute with the name "sn", the "cli_name" for this would be "surename". should i, if present, use "cli_name", instead of "name".

also, some parameters clash with python buildin functions or reserved keywords. i prefixed these cases with "opt_" or "_arg" depending on if they are an option, or an argument in the API description.

this leads to some inconsitencies, example method: user_add, options, "opt_setattr" and "addattr".

i tend to simply prefix all options with either "a_" or "o_"

this way we would get consistent parameter names, like "a_addattr", "a_setattr", which i believe would make the api more easy to understand.

what do you think?

AmbientLighter commented:
Yes, I think it's a good idea to use CLI option name if it exists.

schlitzered committed:
rework meta api method parameter names, to be more consistend
include legacy client.

schlitzered commented:
happy new year,

i just added the following changes:

- strip python_freeipa.client.Client from most API commands
- include legacy client as python_freeipa.client_legacy.ClientLegacy
- include meta client as python_freeipa.client_meta.ClientMeta
- change meta client method argument names to be prefixed with "a_" for "args" and "_o" for "options"

using the CLI option names seemed unreasonable in some cases because of duplicates, and that not all arguments/options have a CLI name.

AmbientLighter commented:
Okay, looks good to me. Could you please, however, fix linter issues and map `Str` type annotation to `str` and `Flag` to `bool`?

AmbientLighter commented on src/python_freeipa/cli_recreate.py:
```diff
@@ -0,0 +1,277 @@
+    source_group = parser.add_mutually_exclusive_group(required=True)
+
+    source_group.add_argument(
+        "--source_json", dest="source_json", action="store", default=None,
```
Usually kebab case is used for argument name. Therefore I would suggest to use `--source-json` for consistency.

schlitzered replied:
changed

AmbientLighter commented on src/python_freeipa/cli_recreate.py:
```diff
@@ -0,0 +1,277 @@
+        help="Full path to json document, describing the API")
+
+    source_group.add_argument(
+        "--source_url", dest="source_url", action="store", default=None,
```
Usually kebab case is used for argument name. Therefore I would suggest to use `--source-url` for consistency.

schlitzered replied:
changed

schlitzered commented:
i have to admit that i actually do not know why tox is complaining there, the syntax is actually valid, and the file can be loaded without any problems.

what do you mean by "map Str type annotation to str and Flag to bool" are you talking about the method descriptions?

AmbientLighter commented:
This code is invalid:

```
  File "/home/travis/build/opennode/python-freeipa/src/python_freeipa/client_meta.py", line 6923
    o_all=True,
        ^
SyntaxError: invalid syntax
```

```
    def env(
            self,
            *args,
            o_all=True,
            o_server=False,
    ):
```

AmbientLighter commented:
> what do you mean by "map Str type annotation to str and Flag to bool" are you talking about the method descriptions?

Yes, that's the case.

AmbientLighter commented:
See also: https://travis-ci.org/opennode/python-freeipa/jobs/633920500

schlitzered commented:
I am sorry, I just checked the "env" method in line 6920, and I fail to see what tox is complaining about. the method works fine, and looks more or less like the others, except that it accepts '*args*. also "pylint" is not complaining about it either.

AmbientLighter commented:
Well, this method does not really accept any unnamed args, according to the freeipa api spec, therefore it makes sense to drop this argument in the first place.

schlitzered commented:
this actually is not true, the "env" method accepts a variable number of arguments, which represent the names of the freippa environment variables that should be returned.

but i think i know what the issue is, it seems like python 2.7 will not accept  *args (or **kwargs) before other named argument names. so i simply have to move it to be the last method.

schlitzered committed:
fix python2 compatibility, adjust method description

schlitzered commented:
i have just adjusted the code to be compatible with python2 *sigh*

i dropped the *args for the "env" method, since there was no obvious way to make this work with python 2. and since this is the only method that behaves like this....

i also adjusted the method documentation as requested, and also put some more info in there in case of Type Int, StrEnum or IntEnum.

let me know what you think

schlitzered committed:
fix tests

schlitzered committed:
fix legacy client

schlitzered committed:
fix some flake8 complaints

schlitzered committed:
fix some flake8 complaints

schlitzered committed:
rename cli arg names

schlitzered committed:
set options, even if empty string, inspired by pullrequest #22

AmbientLighter commented:
@schlitzered Thank you very much for your contribution! It was a pleasure to work with you.

AmbientLighter commented:
I've rolled out release 1.0.0 with meta API.
